crates/wasmtime/src/runtime/component/instance.rs

```diff
@@ -109,46 +195,169 @@ impl Instance {
+            | Export::LiftedFunction { .. }
+            | Export::ModuleStatic { .. }
+            | Export::ModuleImport { .. }
+            | Export::Instance { .. } => None,
+        }
+    }
+
+    /// A methods similar to [`Component::export_index`] except for this
+    /// instance.
+    ///
+    /// This method will lookup the `name` provided within the `instance`
+    /// provided and return a [`ComponentExportIndex`] which can be used to
+    /// pass to other `get_*` functions like [`Instance::get_func`].
     ///
     /// # Panics
     ///
     /// Panics if `store` does not own this instance.
-    pub fn get_resource(&self, mut store: impl AsContextMut, name: &str) -> Option<ResourceType> {
-        self.exports(store.as_context_mut()).root().resource(name)
+    pub fn get_export(
+        &self,
+        mut store: impl AsContextMut,
+        instance: Option<&ComponentExportIndex>,
+        name: &str,
+    ) -> Option<ComponentExportIndex> {
+        self._get_export(store.as_context_mut().0, instance, name)
+    }
+
+    fn _get_export(
+        &self,
+        store: &StoreOpaque,
+        instance: Option<&ComponentExportIndex>,
+        name: &str,
+    ) -> Option<ComponentExportIndex> {
+        let data = store[self.0].as_ref().unwrap();
+        let index = data.component.lookup_export_index(instance, name)?;
+        Some(ComponentExportIndex {
+            id: data.component_id(),
+            index,
+        })
+    }
+
+    fn lookup_export<'a>(
+        &self,
+        store: &'a StoreOpaque,
+        name: impl InstanceExportLookup,
+    ) -> Option<(&'a InstanceData, &'a Export, ExportIndex)> {
+        let data = store[self.0].as_ref().unwrap();
+        let index = name.lookup(&data.component)?;
+        Some((
+            data,
+            &data.component.env_component().export_items[index],
+            index,
+        ))
+    }
+}
+
+/// Trait used to lookup the export of a component instance.
+///
+/// This trait is used as an implementation detail of [`Instance::get_func`]
+/// and related `get_*` methods. Notable implementors of this trait are:
+///
+/// * `str`
+/// * `String`
+/// * [`ComponentExportIndex`]
+///
+/// Note that this is intended to be a `wasmtime`-sealed trait so it shouldn't
+/// need to be implemented externally.
+pub trait InstanceExportLookup {
+    #[doc(hidden)]
+    fn lookup(&self, component: &Component) -> Option<ExportIndex>;
+}
+
+impl<T> InstanceExportLookup for &T
+where
+    T: InstanceExportLookup + ?Sized,
+{
+    fn lookup(&self, component: &Component) -> Option<ExportIndex> {
+        T::lookup(self, component)
+    }
+}
+
+impl InstanceExportLookup for str {
+    fn lookup(&self, component: &Component) -> Option<ExportIndex> {
+        component.env_component().exports.get(self).copied()
+    }
+}
+
+impl InstanceExportLookup for String {
+    fn lookup(&self, component: &Component) -> Option<ExportIndex> {
+        str::lookup(self, component)
```

Comment:
Ah this ends up being a "cute" way to force rustc to do auto-deref without actually doing it yourself. The `str::lookup` function name here is the sugar-y form of `<str as InstanceExportLookup>::lookup` where we're explicitly selecting the impl on `str` so rustc knows ahead of time that the first argument must have type `&str`. Given the `&String` receiver it auto-derefs into `&str`.

(this only works as `str::lookup` isn't a method otherwise or on any other trait)